File: backend/app/api/v1/endpoints/kubernetes.py
from __future__ import annotations

from typing import Annotated, NoReturn

from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.api.deps import get_current_user
from app.core.exceptions import KubernetesClientError, KubernetesResourceNotFoundError
from app.schemas.kubernetes import (
    ClusterSummaryRead,
    DeploymentRead,
    NamespaceRead,
    PodEventRead,
    PodLogRead,
    PodRead,
    ServiceRead,
)
from app.services.kubernetes_service import KubernetesService

router = APIRouter(
    prefix="/kubernetes",
    dependencies=[Depends(get_current_user)],
)


def get_kubernetes_service() -> KubernetesService:
    try:
        return KubernetesService()
    except KubernetesClientError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=str(exc),
        ) from exc


def raise_kubernetes_http_error(exc: KubernetesClientError) -> NoReturn:
    if isinstance(exc, KubernetesResourceNotFoundError):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    raise HTTPException(
        status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
        detail=str(exc),
    ) from exc


@router.get("/summary", response_model=ClusterSummaryRead)
def get_cluster_summary(
    service: Annotated[KubernetesService, Depends(get_kubernetes_service)],
) -> ClusterSummaryRead:
    try:
        return service.get_cluster_summary()
    except KubernetesClientError as exc:
        raise_kubernetes_http_error(exc)


@router.get("/namespaces", response_model=list[NamespaceRead])
def list_namespaces(
    service: Annotated[KubernetesService, Depends(get_kubernetes_service)],
) -> list[NamespaceRead]:
    try:
        return service.list_namespaces()
    except KubernetesClientError as exc:
        raise_kubernetes_http_error(exc)


@router.get("/namespaces/{namespace}/deployments", response_model=list[DeploymentRead])
def list_deployments(
    namespace: str,
    service: Annotated[KubernetesService, Depends(get_kubernetes_service)],
) -> list[DeploymentRead]:
    try:
        return service.list_deployments(namespace)
    except KubernetesClientError as exc:
        raise_kubernetes_http_error(exc)


@router.get("/namespaces/{namespace}/services", response_model=list[ServiceRead])
def list_services(
    namespace: str,
    service: Annotated[KubernetesService, Depends(get_kubernetes_service)],
) -> list[ServiceRead]:
    try:
        return service.list_services(namespace)
    except KubernetesClientError as exc:
        raise_kubernetes_http_error(exc)


@router.get("/namespaces/{namespace}/pods", response_model=list[PodRead])
def list_pods(
    namespace: str,
    service: Annotated[KubernetesService, Depends(get_kubernetes_service)],
) -> list[PodRead]:
    try:
        return service.list_pods(namespace)
    except KubernetesClientError as exc:
        raise_kubernetes_http_error(exc)


@router.get("/namespaces/{namespace}/pods/{pod_name}", response_model=PodRead)
def get_pod(
    namespace: str,
    pod_name: str,
    service: Annotated[KubernetesService, Depends(get_kubernetes_service)],
) -> PodRead:
    try:
        return service.get_pod(namespace, pod_name)
    except KubernetesClientError as exc:
        raise_kubernetes_http_error(exc)


@router.get("/namespaces/{namespace}/pods/{pod_name}/events", response_model=list[PodEventRead])
def list_pod_events(
    namespace: str,
    pod_name: str,
    service: Annotated[KubernetesService, Depends(get_kubernetes_service)],
) -> list[PodEventRead]:
    try:
        return service.list_pod_events(namespace, pod_name)
    except KubernetesClientError as exc:
        raise_kubernetes_http_error(exc)


@router.get("/namespaces/{namespace}/pods/{pod_name}/logs", response_model=PodLogRead)
def get_pod_logs(
    namespace: str,
    pod_name: str,
    service: Annotated[KubernetesService, Depends(get_kubernetes_service)],
    container: str | None = Query(default=None),
    tail_lines: int = Query(default=200, ge=1, le=5000),
    previous: bool = Query(default=False),
    timestamps: bool = Query(default=True),
) -> PodLogRead:
    try:
        return service.get_pod_logs(
            namespace,
            pod_name,
            container=container,
            tail_lines=tail_lines,
            previous=previous,
            timestamps=timestamps,
        )
    except KubernetesClientError as exc:
        raise_kubernetes_http_error(exc)

File: frontend/src/api/kubernetes.ts
import { ApiError, apiRequest, apiUrl, getStoredToken } from "./client";
import type {
  ClusterSummary,
  DeploymentRead,
  NamespaceRead,
  PodEventRead,
  PodHealthAssessmentRead,
  PodLogRead,
  PodRead,
  ServiceRead,
} from "../types/domain";

export function getClusterSummary(): Promise<ClusterSummary> {
  return apiRequest<ClusterSummary>("/api/v1/kubernetes/summary");
}

export function getNamespaces(): Promise<NamespaceRead[]> {
  return apiRequest<NamespaceRead[]>("/api/v1/kubernetes/namespaces");
}

export function getDeployments(namespace: string): Promise<DeploymentRead[]> {
  return apiRequest<DeploymentRead[]>(
    `/api/v1/kubernetes/namespaces/${encodeURIComponent(namespace)}/deployments`,
  );
}

export function getServices(namespace: string): Promise<ServiceRead[]> {
  return apiRequest<ServiceRead[]>(
    `/api/v1/kubernetes/namespaces/${encodeURIComponent(namespace)}/services`,
  );
}

export function getPods(namespace: string): Promise<PodRead[]> {
  return apiRequest<PodRead[]>(
    `/api/v1/kubernetes/namespaces/${encodeURIComponent(namespace)}/pods`,
  );
}

export function getPod(namespace: string, podName: string): Promise<PodRead> {
  return apiRequest<PodRead>(
    `/api/v1/kubernetes/namespaces/${encodeURIComponent(namespace)}/pods/${encodeURIComponent(
      podName,
    )}`,
  );
}

export function getPodEvents(namespace: string, podName: string): Promise<PodEventRead[]> {
  return apiRequest<PodEventRead[]>(
    `/api/v1/kubernetes/namespaces/${encodeURIComponent(namespace)}/pods/${encodeURIComponent(
      podName,
    )}/events`,
  );
}

export function getPodLogs(namespace: string, podName: string): Promise<PodLogRead> {
  return apiRequest<PodLogRead>(
    `/api/v1/kubernetes/namespaces/${encodeURIComponent(namespace)}/pods/${encodeURIComponent(
      podName,
    )}/logs?tail_lines=300&timestamps=true`,
  );
}

interface StreamPodLogsOptions {
  tailLines?: number;
  timestamps?: boolean;
  previous?: boolean;
  container?: string | null;
}

interface StreamPodLogsCallbacks {
  onLine: (line: string) => void;
  onError?: (message: string) => void;
}

export async function streamPodLogs(
  namespace: string,
  podName: string,
  options: StreamPodLogsOptions,
  signal: AbortSignal,
  callbacks: StreamPodLogsCallbacks,
): Promise<void> {
  const params = new URLSearchParams({
    tail_lines: String(options.tailLines ?? 100),
    timestamps: String(options.timestamps ?? true),
    previous: String(options.previous ?? false),
  });
  if (options.container) {
    params.set("container", options.container);
  }

  const headers = new Headers();
  const token = getStoredToken();
  if (token) {
    headers.set("Authorization", `Bearer ${token}`);
  }

  const response = await fetch(
    apiUrl(
      `/api/v1/kubernetes/namespaces/${encodeURIComponent(namespace)}/pods/${encodeURIComponent(
        podName,
      )}/logs/stream?${params.toString()}`,
    ),
    { headers, signal },
  );

  if (!response.ok) {
    throw new ApiError(response.statusText || "Unable to stream logs", response.status);
  }
  if (!response.body) {
    throw new Error("Log stream is unavailable in this browser.");
  }

  await readServerSentEvents(response.body, callbacks, signal);
}

async function readServerSentEvents(
  body: ReadableStream<Uint8Array>,
  callbacks: StreamPodLogsCallbacks,
  signal: AbortSignal,
): Promise<void> {
  const reader = body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  try {
    while (!signal.aborted) {
      const { done, value } = await reader.read();
      if (done) {
        break;
      }
      buffer += decoder.decode(value, { stream: true });
      const events = buffer.split("\n\n");
      buffer = events.pop() ?? "";
      for (const event of events) {
        handleServerSentEvent(event, callbacks);
      }
    }
  } finally {
    reader.releaseLock();
  }
}

function handleServerSentEvent(event: string, callbacks: StreamPodLogsCallbacks): void {
  const lines = event.split("\n");
  const eventType = lines.find((line) => line.startsWith("event:"))?.slice(6).trim();
  const data = lines
    .filter((line) => line.startsWith("data:"))
    .map((line) => line.slice(5).trimStart())
    .join("\n");
  if (!data) {
    return;
  }

  const payload = JSON.parse(data) as { line?: string; detail?: string };
  if (eventType === "error") {
    callbacks.onError?.(payload.detail ?? "Log stream failed.");
    return;
  }
  if (payload.line !== undefined) {
    callbacks.onLine(payload.line);
  }
}

export function getPodHealth(
  namespace: string,
  podName: string,
): Promise<PodHealthAssessmentRead> {
  return apiRequest<PodHealthAssessmentRead>(
    `/api/v1/monitoring/namespaces/${encodeURIComponent(namespace)}/pods/${encodeURIComponent(
      podName,
    )}/health?include_logs=true`,
  );
}
